### PR TITLE
Dev piotrek servertests

### PR DESCRIPTION
**Pull request by sabath91**

### CODE REVIEW

File: client/src/main/java/com/epam/ships/client/client/Client.java [hunk 1]
```diff
@@ -42,35 +42,31 @@ public boolean connect(final String ipAddress, final int port) {
 
     @Override
     public void run() {
-        listenLoop();
+        try {
+            listenLoop();
+        } catch (IOException e) {
```
Comment (korotkevics) [inline client/src/main/java/com/epam/ships/client/client/Client.java]: final

File: server/src/main/java/com/epam/ships/server/CommunicationBus.java [hunk 2]
```diff
@@ -50,11 +50,19 @@ void stop() {
...
 
-    Message receive() {
-        return this.clients.get(0).receive();
+    Message receive(WrappedClient sender) {
+        return sender.receive();
     }
 
-    public void send(Message response) {
-        this.clients.get(0).send(response);
+    void send(WrappedClient recipient, Message response) {
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/CommunicationBus.java]: final to parameters

File: server/src/main/java/com/epam/ships/server/CommunicationBus.java [hunk 3]
```diff
@@ -50,11 +50,19 @@ void stop() {
...
+        recipient.send(response);
+    }
+
+    WrappedClient getFirstClient(){
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/CommunicationBus.java]: final to returning type

File: server/src/main/java/com/epam/ships/server/CommunicationBus.java [hunk 4]
```diff
@@ -50,11 +50,19 @@ void stop() {
...
+        return clients.get(0);
+    }
+
+    WrappedClient getSecondClient(){
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/CommunicationBus.java]: final to returning type

File: server/src/main/java/com/epam/ships/server/CommunicationBus.java [hunk 5]
```diff
@@ -50,11 +50,19 @@ void stop() {
...
     }
 
-    public void send(Message response) {
-        this.clients.get(0).send(response);
+    void send(WrappedClient recipient, Message response) {
+        recipient.send(response);
+    }
+
+    WrappedClient getFirstClient(){
+        return clients.get(0);
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/CommunicationBus.java]: this

File: server/src/main/java/com/epam/ships/server/CommunicationBus.java [hunk 6]
```diff
@@ -50,11 +50,19 @@ void stop() {
...
+    }
+
+    WrappedClient getSecondClient(){
+        return clients.get(1);
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/CommunicationBus.java]: this

File: server/src/main/java/com/epam/ships/server/Game.java [hunk 7]
```diff
@@ -17,54 +17,56 @@
 
     private CommunicationBus communicationBus;
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Game.java]: final

File: server/src/main/java/com/epam/ships/server/Game.java [hunk 8]
```diff
@@ -17,54 +17,56 @@
...
+    void play() {
+        notifyPlayersThatTheyCanStartGame();
+//        TODO: notify Users both are ready
+        boolean isGameFinished = false;
+        while (!isGameFinished ){
+            exchangeGreetings();
+            turnManager.switchPlayer();
+//            TODO: game loop until we will find winner
+//            isGameFinished = some referee method?
+            waitAWhile();
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Game.java]: rest()?

File: server/src/main/java/com/epam/ships/server/Game.java [hunk 9]
```diff
@@ -17,54 +17,56 @@
...
+        }
+    }
 
-                MessageBuilder messageBuilder = new MessageBuilder()
-                        .withAuthor("Name Validation")
-                        .withHeader("Server");
 
-                Message message;
+    private void exchangeGreetings(){
+        MessageBuilder messageBuilder = new MessageBuilder();
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Game.java]: final

File: server/src/main/java/com/epam/ships/server/Game.java [hunk 10]
```diff
@@ -17,54 +17,56 @@
...
+        Message message = messageBuilder.withAuthor("server").withHeader("greetings").withStatus("OK").withStatement("Welcome on board").build();
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Game.java]: Use chained call

File: server/src/main/java/com/epam/ships/server/Game.java [hunk 11]
```diff
@@ -17,54 +17,56 @@
...
+        communicationBus.send(turnManager.getCurrentPlayer(), message);
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Game.java]: this

File: server/src/main/java/com/epam/ships/server/Game.java [hunk 12]
```diff
@@ -17,54 +17,56 @@
...
+        Message greetings = communicationBus.receive(turnManager.getCurrentPlayer());
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Game.java]: final

File: server/src/main/java/com/epam/ships/server/Game.java [hunk 13]
```diff
@@ -17,54 +17,56 @@
...
-                            .withStatement("No warnings")
-                            .build();
-                } else {
-                    message = messageBuilder
-                            .withStatus("ERR")
-                            .withStatement("Name should begin with a capital letter")
-                            .build();
-                }
+    private void notifyPlayersThatTheyCanStartGame(){
+       opponentConnected();
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Game.java]: this

File: server/src/main/java/com/epam/ships/server/Game.java [hunk 14]
```diff
@@ -17,54 +17,56 @@
...
+       turnManager.switchPlayer();
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Game.java]: this

File: server/src/main/java/com/epam/ships/server/Game.java [hunk 15]
```diff
@@ -17,54 +17,56 @@
...
+       opponentConnected();
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Game.java]: this

File: server/src/main/java/com/epam/ships/server/Game.java [hunk 16]
```diff
@@ -17,54 +17,56 @@
...
+   }
 
-                communicationBus.send(message);
-            }
+    private void opponentConnected(){
+        MessageBuilder messageBuilder = new MessageBuilder();
+        Message message = messageBuilder.withAuthor("server").withHeader("opponentConnected").withStatus("OK").withStatement("true").build();
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Game.java]: Merge with MessageBuilder with a chained call and use final

File: server/src/main/java/com/epam/ships/server/Game.java [hunk 17]
```diff
@@ -17,54 +17,56 @@
...
+        communicationBus.send(turnManager.getCurrentPlayer(), message);
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Game.java]: this

File: server/src/main/java/com/epam/ships/server/Game.java [hunk 18]
```diff
@@ -17,54 +17,56 @@
...
-            if ("Connection".equals(messageReceived.getHeader()) && "END".equals(messageReceived.getStatus())) {
-                flag = false;
-            }
-            try {
-                Thread.sleep(250);
-            } catch (InterruptedException e) {
-                logger.error(e.getMessage());
-            }
+    private void waitAWhile() {
+        final long aWhile = 300;
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Game.java]: waitingTime

File: server/src/main/java/com/epam/ships/server/Game.java [hunk 19]
```diff
@@ -17,54 +17,56 @@
...
+        try {
+            Thread.sleep(aWhile);
+        } catch (InterruptedException e) {
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Game.java]: final

File: server/src/main/java/com/epam/ships/server/Main.java [hunk 20]
```diff
@@ -15,7 +15,7 @@ public static void main(String[] args) throws IOException {
         while (true) {
             communicationBus.start();
             Game game = new Game(communicationBus);
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Main.java]: final

File: server/src/main/java/com/epam/ships/server/Main.java [hunk 21]
```diff
@@ -15,7 +15,7 @@ public static void main(String[] args) throws IOException {
...
-            game.letsChatALittleBit();
+            game.play();
             try {
                 Thread.sleep(300);
             } catch (InterruptedException e) {
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/Main.java]: final

File: server/src/main/java/com/epam/ships/server/TurnManager.java [hunk 22]
```diff
@@ -0,0 +1,25 @@
+package com.epam.ships.server;
+
+class TurnManager {
+    private final WrappedClient firstPlayer;
+    private final WrappedClient secondPlayer;
+    private WrappedClient currentPlayer;
+
+    TurnManager(WrappedClient firstPlayer, WrappedClient secondPlayer) {
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/TurnManager.java]: final to arguments

File: server/src/main/java/com/epam/ships/server/TurnManager.java [hunk 23]
```diff
@@ -0,0 +1,25 @@
+package com.epam.ships.server;
+
+class TurnManager {
+    private final WrappedClient firstPlayer;
+    private final WrappedClient secondPlayer;
+    private WrappedClient currentPlayer;
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/TurnManager.java]: final

File: server/src/main/java/com/epam/ships/server/TurnManager.java [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+
+    TurnManager(WrappedClient firstPlayer, WrappedClient secondPlayer) {
+        this.firstPlayer = firstPlayer;
+        this.secondPlayer = secondPlayer;
+        this.currentPlayer = firstPlayer;
+    }
+
+    WrappedClient getCurrentPlayer() {
+        return currentPlayer;
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/TurnManager.java]: this

File: server/src/main/java/com/epam/ships/server/TurnManager.java [hunk 25]
```diff
@@ -0,0 +1,25 @@
...
+    private final WrappedClient secondPlayer;
+    private WrappedClient currentPlayer;
+
+    TurnManager(WrappedClient firstPlayer, WrappedClient secondPlayer) {
+        this.firstPlayer = firstPlayer;
+        this.secondPlayer = secondPlayer;
+        this.currentPlayer = firstPlayer;
+    }
+
+    WrappedClient getCurrentPlayer() {
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/TurnManager.java]: final

File: server/src/main/java/com/epam/ships/server/TurnManager.java [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+        return currentPlayer;
+    }
+
+    void switchPlayer() {
+        if (currentPlayer.equals(firstPlayer)) {
+            currentPlayer = secondPlayer;
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/TurnManager.java]: this

File: server/src/main/java/com/epam/ships/server/TurnManager.java [hunk 27]
```diff
@@ -0,0 +1,25 @@
...
+        } else {
+            currentPlayer = firstPlayer;
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/TurnManager.java]: this

File: server/src/main/java/com/epam/ships/server/WrappedClient.java [hunk 28]
```diff
@@ -51,7 +51,7 @@ void close() {
         }
     }
 
-    public void send(Message message) {
+    void send(Message message) {
```
Comment (korotkevics) [inline server/src/main/java/com/epam/ships/server/WrappedClient.java]: final

File: server/src/main/java/com/epam/ships/server/TurnManager.java (hunk 26 above)
Comment (korotkevics) [reply server/src/main/java/com/epam/ships/server/TurnManager.java]: Add hashcode + equals for Player

File: server/src/test/java/com/epam/ships/server/AppServerTest.java [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.net.Socket;
+
+import static org.testng.Assert.assertEquals;
+
+@Test
+public class AppServerTest {
```
Comment (korotkevics) [inline server/src/test/java/com/epam/ships/server/AppServerTest.java]: Remove empty lines

**Comment by sabath91** [comment]
All done

**Pull request merged**